homeassistant/components/thermostat/ecobee.py

```diff
@@ -0,0 +1,263 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ECOBEE_CONFIG_FILE = 'ecobee.conf'
+_CONFIGURING = {}
+
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """ Setup Platform """
+    # Only act if we are not already configuring this host
+    if 'ecobee' in _CONFIGURING:
+        return
+
+    setup_ecobee(hass, config, add_devices_callback)
+
+
+def setup_ecobee(hass, config, add_devices_callback):
+    """ Setup ecobee thermostat """
+    from pyecobee import Ecobee, config_from_file
+    # Create ecobee.conf if it doesn't exist
+    if not os.path.isfile(hass.config.path(ECOBEE_CONFIG_FILE)):
+        jsonconfig = {"API_KEY": config[CONF_API_KEY]}
+        config_from_file(hass.config.path(ECOBEE_CONFIG_FILE), jsonconfig)
+
+    ecobee = Ecobee(hass.config.path(ECOBEE_CONFIG_FILE))
+
+    # If ecobee has a PIN then it needs to be configured.
+    if ecobee.pin is not None:
+        # ecobee.request_pin()
+        request_configuration(ecobee, hass, add_devices_callback)
+        return
+
+    if 'ecobee' in _CONFIGURING:
+        _CONFIGURING.pop('ecobee')
+        configurator = get_component('configurator')
+        configurator.request_done('ecobee')
+
+    devices = []
+    for index in range(0, len(ecobee.thermostats)):
+        devices.append(Thermostat(ecobee, index))
+
+    add_devices_callback(devices)
+
+
+def request_configuration(ecobee, hass, add_devices_callback):
+    """ Request configuration steps from the user. """
+    configurator = get_component('configurator')
+    if 'ecobee' in _CONFIGURING:
+        configurator.notify_errors(
+            _CONFIGURING['ecobee'], "Failed to register, please try again.")
+
+        return
+
+    # pylint: disable=unused-argument
+    def ecobee_configuration_callback(data):
+        """ Actions to do when our configuration callback is called. """
+        ecobee.request_tokens()
+        ecobee.update()
+        setup_ecobee(hass, None, add_devices_callback)
+
+    _CONFIGURING['ecobee'] = configurator.request_config(
+        hass, "Ecobee", ecobee_configuration_callback,
+        description=(
+            'Please authorize this app at https://www.ecobee.com/consumer'
+            'portal/index.html with pin code: ' + ecobee.pin),
+        description_image='https://goo.gl/6tBkbH',
+        submit_caption="I have authorized the app."
+    )
+
+
+class Thermostat(ThermostatDevice):
+    """docstring for Thermostat"""
+
+    def __init__(self, ecobee, thermostat_index):
+        self.ecobee = ecobee
+        self.thermostat_index = thermostat_index
+        self.thermostat_data = self.ecobee.get_thermostat(
+            self.thermostat_index)
+        self._name = self.thermostat_data['name']
+        if 'away' in self.thermostat_data['program']['currentClimateRef']:
+            self._away = True
+        else:
+            self._away = False
+
+    def update(self):
+        self.thermostat_data = self.ecobee.get_thermostat(
```

Comment:
This method can potentially be very expensive because each call to `get_thermostat` results in a full refresh from the API.

It also seems to me that the sensor values are not being updated in the config. I think it would make more sense for the sensors to instantiate their own ecobee API object (using the config file) and query for sensor data using `ecobee.get_remote_sensors()`. To prevent from over-querying, consider using a similar setup as the [Bitcoin sensor](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/sensor/bitcoin.py#L179-L193) which wraps the update method with a throttle and caches the result.